query/src/sql/optimizer/cascades/mod.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod explore_rules;
+mod implement_rules;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::sql::optimizer::cascades::explore_rules::get_explore_rule_set;
+use crate::sql::optimizer::cascades::implement_rules::get_implement_rule_set;
+use crate::sql::optimizer::group::Group;
+use crate::sql::optimizer::m_expr::MExpr;
+use crate::sql::optimizer::memo::Memo;
+use crate::sql::optimizer::optimize_context::OptimizeContext;
+use crate::sql::optimizer::rule::RulePtr;
+use crate::sql::optimizer::rule::RuleSet;
+use crate::sql::optimizer::rule::TransformState;
+use crate::sql::optimizer::PhysicalProperty;
+use crate::sql::optimizer::RequiredProperty;
+use crate::sql::optimizer::SExpr;
+use crate::sql::IndexType;
+
+/// A cascades-style search engine to enumerate possible alternations of a relational expression and
+/// find the optimal one.
+///
+/// NOTICE: we don't support cost-based optimization and lower bound searching for now.
+pub struct CascadesOptimizer {
+    optimize_context: OptimizeContext,
+    memo: Memo,
+    explore_rules: RuleSet,
+    implement_rules: RuleSet,
+}
+
+impl CascadesOptimizer {
+    pub fn create(optimize_context: OptimizeContext) -> Self {
+        CascadesOptimizer {
+            optimize_context,
+            memo: Memo::create(),
+            explore_rules: get_explore_rule_set(),
+            implement_rules: get_implement_rule_set(),
+        }
+    }
+
+    fn init(&mut self, expression: SExpr) -> Result<()> {
+        self.memo.init(expression)?;
+
+        Ok(())
+    }
+
+    pub fn optimize(&mut self, expression: SExpr) -> Result<SExpr> {
+        self.init(expression)?;
+
+        self.explore_group(self.memo.root().unwrap().group_index())?;
+
+        self.implement_group(self.memo.root().unwrap().group_index())?;
+
+        self.find_optimal_plan()
+    }
+
+    fn explore_group(&mut self, group_index: IndexType) -> Result<()> {
```

Comment:
Maybe should add a todo comment. we need to use while to rewrite the 
 explore task(maybe also other task)?

Reply:
You remind me about this. I was going to unify explore and implement transform task but forgot to do that.